### PR TITLE
ReCoRD

### PR DESCRIPTION
**Pull request by W4ngatang**
Implements ReCoRD task.

We model the task as a multiple choice question answering task where answer choices are given in the data and the questions are Cloze-style. Due to memory constraints, currently we substitute each answer choice into the question and treat it as its own example. For validation / test, we aggregate the probabilities of all answer choices for a question, and take the most probable choice as the model's prediction.

### CODE REVIEW

File: src/tasks/qa.py [hunk 1]
```diff
@@ -172,3 +191,188 @@ def get_metrics(self, reset=False):
...
+                     }
+                examples.append(ex)
+
+        return examples
+
+    def _load_answers(self) -> None:
+        """ """
+        answers = {}
+        for split, split_path in self.files_by_split.items():
+            #if split == "test":
```
Comment (W4ngatang) [inline src/tasks/qa.py]: to add back in

File: scripts/superglue-baselines.sh [hunk 2]
```diff
@@ -18,7 +18,11 @@ function copa() {
 }
 
 function multirc() {
-    python main.py --config config/superglue-bert.conf --overrides "random_seed = ${seed}, cuda = ${gpuid}, run_name = multirc, pretrain_tasks = \"multirc\", target_tasks = \"multirc\", do_pretrain = 1, do_target_task_training = 0, do_full_eval = 1, batch_size = 4, val_interval = 1000"
+    python main.py --config config/superglue-bert.conf --overrides "random_seed = ${seed}, cuda = ${gpuid}, run_name = multirc, pretrain_tasks = \"multirc\", target_tasks = \"multirc\", do_pretrain = 1, do_target_task_training = 0, do_full_eval = 1, batch_size = 4, val_interval = 1000, val_data_limit = -1"
+}
+
+function record() {
+    python -m ipdb main.py --config config/superglue-bert.conf --overrides "random_seed = ${seed}, cuda = ${gpuid}, run_name = record, pretrain_tasks = \"record\", target_tasks = \"record\", do_pretrain = 1, do_target_task_training = 0, do_full_eval = 1, batch_size = 8, val_interval = 10000, val_data_limit = -1"
```
Comment (W4ngatang) [inline scripts/superglue-baselines.sh]: whoops, should delete `ipdb`

File: main.py [hunk 3]
```diff
@@ -281,7 +281,7 @@ def evaluate_and_write(args, model, tasks, splits_to_write):
             tasks, val_preds, args.run_dir, "val", strict_glue_format=args.write_strict_glue_format
         )
     if "test" in splits_to_write:
-        _, te_preds = evaluate.evaluate(model, tasks, args.batch_size, args.cuda, "test")
+        te_results, te_preds = evaluate.evaluate(model, tasks, args.batch_size, args.cuda, "test")
```
Comment (W4ngatang) [inline main.py]: remove

**Comment by W4ngatang** [comment]
Waiting on verifying with Amanpreet that our numbers match. Also need to rerun MultiRC to make sure I didn't break stuff.

File: scripts/superglue-baselines.sh [hunk 4]
```diff
@@ -18,7 +18,11 @@ function copa() {
 }
 
 function multirc() {
-    python main.py --config config/superglue-bert.conf --overrides "random_seed = ${seed}, cuda = ${gpuid}, run_name = multirc, pretrain_tasks = \"multirc\", target_tasks = \"multirc\", do_pretrain = 1, do_target_task_training = 0, do_full_eval = 1, batch_size = 4, val_interval = 1000"
+    python main.py --config config/superglue-bert.conf --overrides "random_seed = ${seed}, cuda = ${gpuid}, run_name = multirc, pretrain_tasks = \"multirc\", target_tasks = \"multirc\", do_pretrain = 1, do_target_task_training = 0, do_full_eval = 1, batch_size = 4, val_interval = 1000, val_data_limit = -1"
```
Comment (sleepinyourhat) [inline scripts/superglue-baselines.sh]: is 5k (default val data limit) vs. 10k really crucial during training?

File: src/models.py [hunk 5]
```diff
@@ -1158,30 +1158,47 @@ def _multiple_choice_reading_comprehension_forward(self, batch, task, predict):
...
             # else, we embed each independently and concat them
-            par_emb, par_mask = self.sent_encoder(batch["paragraph"], task)
-            qst_emb, qst_mask = self.sent_encoder(batch["question"], task)
-            ans_emb, ans_mask = self.sent_encoder(batch["answer"], task)
-            inp = torch.cat([par_emb, qst_emb, ans_emb], dim=1)
-            inp_mask = torch.cat([par_mask, qst_mask, ans_mask], dim=1)
+            psg_emb, psg_mask = self.sent_encoder(batch["psg"], task)
+            qst_emb, qst_mask = self.sent_encoder(batch["qst"], task)
+
+            if "ans" in batch:
```
Comment (sleepinyourhat) [inline src/models.py]: Add a comment to spell out this logic a bit more clearly.

File: src/tasks/qa.py [hunk 6]
```diff
@@ -17,6 +17,27 @@
 from .registry import register_task
 
 
+def _get_f1(x, y):
```
Comment (sleepinyourhat) [inline src/tasks/qa.py]: Why do we need this? Add a header comment.

File: src/tasks/qa.py [hunk 7]
```diff
@@ -172,3 +193,191 @@ def get_metrics(self, reset=False):
...
+        Split should be one of "train", "val", or "test".
+        """
+        return self.load_data_for_path(self.files_by_split[split], split)
+
+    def load_data_for_path(self, path, split):
+        """ Load data """
+
+        def split_then_tokenize(sent):
+            """ Tokenize questions while preserving @placeholder token """
+            sent_parts = sent.split("@placeholder")
```
Comment (sleepinyourhat) [inline src/tasks/qa.py]: What is this?

File: src/tasks/qa.py [hunk 8]
```diff
@@ -172,3 +193,191 @@ def get_metrics(self, reset=False):
...
+    def process_split(self, split, indexers) -> Iterable[Type[Instance]]:
+        """ Process split text into a list of AllenNLP Instances. """
+        is_using_bert = "bert_wpm_pretokenized" in indexers
+
+        def is_answer(x, ys):
+            """ Given a list of answers, determine if x is an answer """
+            return x in ys
+
+        def insert_ent(ent, template):
+            """ Replace ent into template """
```
Comment (sleepinyourhat) [inline src/tasks/qa.py]: ?

File: src/tasks/qa.py [hunk 9]
```diff
@@ -172,3 +193,191 @@ def get_metrics(self, reset=False):
...
+            anss = example["answers"]
+            par_idx = example["psg_id"]
+            qst_idx = example["qst_id"]
+            for ent_idx, (ent, ent_str) in enumerate(zip(ents, ent_strs)):
+                label = is_answer(ent_str, anss)
+                qst = insert_ent(ent, qst_template)
+                yield _make_instance(psg, qst, ent_str, label, par_idx, qst_idx, ent_idx)
+
+    def count_examples(self):
+        """ Compute here b/c we"re streaming the sentences. """
```
Comment (sleepinyourhat) [inline src/tasks/qa.py]: we're

File: scripts/superglue-baselines.sh (hunk 4 above)
Comment (W4ngatang) [reply scripts/superglue-baselines.sh]: Strictly speaking, no, but it's useful. Evaluation/validation takes a while because there is a lot of data and 5k batches is only ~1% of the training data, so this greatly speeds things up.

File: src/tasks/qa.py (hunk 6 above)
Comment (W4ngatang) [reply src/tasks/qa.py]: ReCoRD's metrics involve computing token-level F1 between a prediction and gold answer. This is my version, but I'm adding in the function from the official eval script.

File: src/tasks/qa.py (hunk 7 above)
Comment (W4ngatang) [reply src/tasks/qa.py]: ReCoRD's queries are Cloze style, so we want to tokenize while preserving the special placeholder token that indicates the masked out word.

File: src/tasks/qa.py (hunk 8 above)
Comment (W4ngatang) [reply src/tasks/qa.py]: ReCoRD's queries are Cloze style, so this inserts a possible answer candidate into the "query".

File: scripts/superglue-baselines.sh (hunk 4 above)
Comment (W4ngatang) [reply scripts/superglue-baselines.sh]: (note: this comment applies below)

**Comment by W4ngatang** [comment]
Currently waiting on: 
- verifying numbers computed by code match eval server
- testing MultiRC
- removing extra comments and functions.
